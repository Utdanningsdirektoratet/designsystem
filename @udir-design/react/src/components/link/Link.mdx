import { Meta, Canvas, Controls, Primary } from '@storybook/addon-docs/blocks';
import * as LinkStories from './Link.stories';

<Meta of={LinkStories} />

# Link

`Link` er en klikkbar tekst,
som gjør det enkelt og raskt for brukerne å navigere til filer eller sider internt eller eksternt.
Komponenten er basert på [anchor HTML-taggen](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).

**Passer til å**

- lenke til en annen side
- åpne en fil
- navigere til en annen del av samme side

**Passer ikke til å**

- utføre handlinger

## Slik bruker du Link

<Primary />
<Controls />

```tsx
import { Link } from '@udir-design/react';

<Link href="https://udir.no/">Gå til udir.no</Link>;
```

## Eksempler

### I tekst

En lenke brukes ofte som en del av en tekst. Teksten som omfattes av lenken bør indikere tydelig hvor den går.

<Canvas of={LinkStories.InText} />

### Med ikon

Ikonet legges på venstre side. Dette er et funksjonelt ikon som forteller brukerne hva lenken leder dem til.
Ikonet burde være skjult fra skjermleser, ved å legge på `aria-hidden`.
Plasser teksten i `<span>` for å få anbefalt mellomrom uten understrek mellom ikon og tekst.

<Canvas of={LinkStories.WithIcon} />

### Flere linjer

<Canvas of={LinkStories.LongLink} />

### Åpne fil

<Canvas of={LinkStories.File} />

## Retningslinjer

### Åpne lenker i samme fane/vindu

Både interne og eksterne lenker bør åpnes i samme fane/vindu. Da sikrer vi en forutsigbar brukeropplevelse fordi brukerens egne lenkeinnstillinger i nettleseren videreføres. Flere bruker Tilbake-knappen i nettleseren hvis lenken ikke førte dit de trodde, men det fungerer ikke når lenken ble åpnet i en ny fane.

Det finnes noen unntak der det kan være hensiktsmessig å åpne lenken i en ny fane. Det kan for eksempel være:

- Når brukeren er midt i en prosess, for eksempel ved utfylling av et skjema.
- Når lenken leder fra en innlogget side til en åpen side.
- Når brukeren har behov for å sammenligne innhold på to sider.

Det skal da informeres om at lenken åpnes i en ny fane.

### Hvor fører lenken

Det er ikke alltid nødvendig å merke at en lenke er ekstern. Ofte fremgår det tydelig av konteksten. Hvis lenketeksten for eksempel er “Flekkefjord kommune”, vil brukeren naturlig forvente at lenken leder til kommunens nettsider. I slike tilfeller er det unødvendig å legge til ekstra informasjon.

Dersom lenken viser til et spesifikt innhold hos kommunen, som for eksempel informasjon om parkering, kan lenken være "Parkering i Flekkefjord kommune" eller "Parkering (flekkefjord.kommune.no)". Det viktigste er at brukeren ikke blir overrasket over hvor de havner etter å ha klikket.

### Lenke til filer

Du må informere brukere når lenken leder til et dokument (PDF, Excel, osv.). Det er i tillegg god praksis å informere brukeren om dokumentets størrelse. Det enkleste er å skrive informasjonen i selv lenken. Inkluder gjerne også et ikon.

## Tekst i Link

En god lenketekst gjør det lettere for folk å bestemme om de ønsker å klikke på lenken eller ikke. Det kan være særlig kostbart for de med nedsatt funksjonsevne å følge en lenke til noe uforventet.
Det kan også koste tid og penger for de med dårlig eller dyr nettforbindelse.

Følg disse retningslinjene for å lage gode og selvforklarende lenker:

- Bruk beskrivende og forklarende lenketekster.
- Skriv korte lenketekster. Ikke prøv å gjenskape informasjon på siden du lenker til, bare si hva innholdet dreier seg om.
- Start lenketeksten med den mest relevante informasjonen. Det gjør det lettere for brukerne å skanne teksten og finne det de leter etter.
- Pass på at du ikke har flere lenker med lik lenketekst på samme side, med mindre de lenker til samme sted.

Vi bruker aldri lenketekster som “Klikk her”, “Se her” og “Les mer”. Det er fordi brukerne ofte skanner teksten og leter etter veier videre. Når brukerne kan lese et meningsbærende innhold i lenketeksten, finner de raskt det innholdet de leter etter.

[Les mer utfyllende om hvordan lage gode lenker på designsystemet.no](https://www.designsystemet.no/no/patterns/external-links)

## Tilgjengelighet

`Link` fungerer som et [`Anchor element`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/a).

Når lenken får fokus gjennom tastaturnavigasjon får den bakgrunnsfarge.
Lenker som er besøkt vil ha en annen farge enn lenker som ikke er besøkt. Dette er for å hjelpe brukerne med å forstå hvilke lenker de allerede har klikket på.

Dersom du bruker bare et bilde som en lenke, må bildet være kodet med alternativ tekst, ellers så vet ikke skjermlesere eller søkemotorer hvor lenken leder til.

Dersom det finnes en lenketekst rett ved siden av bildet, skal både teksten og bildet være wrappet i én anchor-tag, og bildet skal ha tom alt-attributt (alt="").
