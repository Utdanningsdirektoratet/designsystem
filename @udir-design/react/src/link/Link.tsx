export { Link, type LinkProps } from '@digdir/designsystemet-react';
