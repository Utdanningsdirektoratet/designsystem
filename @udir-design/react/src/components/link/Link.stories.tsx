import type { Meta, StoryObj } from '@storybook/react';
import { Link } from './Link';

const meta: Meta<typeof Link> = {
  component: Link,
  tags: ['alpha'],
};

export default meta;
type Story = StoryObj<typeof Link>;

export const Preview: Story = {
  args: {
    href: 'https://www.udir.no/',
    children: 'GÃ¥ til udir.no',
  },
};
