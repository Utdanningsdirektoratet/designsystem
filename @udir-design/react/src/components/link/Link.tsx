import {
  Link as DigdirLink,
  type LinkProps as DigdirLinkProps,
} from '@digdir/designsystemet-react';
import { ComponentRef, ForwardRefExoticComponent, RefAttributes } from 'react';
import './link.css';

type LinkProps = Omit<DigdirLinkProps, 'data-color'>;

const Link = DigdirLink as ForwardRefExoticComponent<
  LinkProps & RefAttributes<ComponentRef<typeof DigdirLink>>
>;

export { Link, LinkProps };
