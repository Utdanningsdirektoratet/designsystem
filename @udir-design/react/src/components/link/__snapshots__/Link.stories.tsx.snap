// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Anchor Links 1`] = `
<div
  data-variant="tinted"
  data-color="accent"
>
  <h2 data-size="xs">
    Innhold på siden
  </h2>
  <ul style="list-style: none; padding: 0px; margin-top: var(--ds-size-4);">
    <li>
      <a href="#retningslinjer">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          fill="none"
          viewbox="0 0 24 24"
          focusable="false"
          role="img"
          aria-hidden="true"
        >
          <path
            fill="currentColor"
            d="M6.5 11.25c-.69 0-1.25-.56-1.25-1.25V4a.75.75 0 0 0-1.5 0v6a2.75 2.75 0 0 0 2.75 2.75h10.69l-3.22 3.22a.75.75 0 1 0 1.06 1.06l4.5-4.5a.75.75 0 0 0 0-1.06l-4.5-4.5a.75.75 0 0 0-1.06 1.06l3.22 3.22z"
          >
          </path>
        </svg>
        Retningslinjer
      </a>
    </li>
    <li>
      <a href="#tekst-i-link">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          fill="none"
          viewbox="0 0 24 24"
          focusable="false"
          role="img"
          aria-hidden="true"
        >
          <path
            fill="currentColor"
            d="M6.5 11.25c-.69 0-1.25-.56-1.25-1.25V4a.75.75 0 0 0-1.5 0v6a2.75 2.75 0 0 0 2.75 2.75h10.69l-3.22 3.22a.75.75 0 1 0 1.06 1.06l4.5-4.5a.75.75 0 0 0 0-1.06l-4.5-4.5a.75.75 0 0 0-1.06 1.06l3.22 3.22z"
          >
          </path>
        </svg>
        Tekst i Link
      </a>
    </li>
    <li>
      <a href="#tilgjengelighet">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="1em"
          height="1em"
          fill="none"
          viewbox="0 0 24 24"
          focusable="false"
          role="img"
          aria-hidden="true"
        >
          <path
            fill="currentColor"
            d="M6.5 11.25c-.69 0-1.25-.56-1.25-1.25V4a.75.75 0 0 0-1.5 0v6a2.75 2.75 0 0 0 2.75 2.75h10.69l-3.22 3.22a.75.75 0 1 0 1.06 1.06l4.5-4.5a.75.75 0 0 0 0-1.06l-4.5-4.5a.75.75 0 0 0-1.06 1.06l3.22 3.22z"
          >
          </path>
        </svg>
        Tilgjengelighet
      </a>
    </li>
  </ul>
</div>
`;

exports[`File 1`] = `
<a href="https://www.udir.no/api/PdfApi/PrintPageAsPdfDocument/224209">
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="1em"
    height="1em"
    fill="none"
    viewbox="0 0 24 24"
    focusable="false"
    role="img"
    aria-hidden="true"
  >
    <path
      fill="currentColor"
      fill-rule="evenodd"
      d="M13.25 3.5a.25.25 0 0 0-.25-.25H6.5c-.69 0-1.25.56-1.25 1.25v6c0 .138.112.25.25.25h13a.25.25 0 0 0 .25-.25V9a.25.25 0 0 0-.25-.25h-4c-.69 0-1.25-.56-1.25-1.25zM15 7.25a.25.25 0 0 1-.25-.25V4.293a.25.25 0 0 1 .427-.177l2.707 2.707a.25.25 0 0 1-.177.427zm-12 5a.75.75 0 0 0-.75.75v7c0 .414.336.75.75.75h18a.75.75 0 0 0 .75-.75v-7a.75.75 0 0 0-.75-.75zm4.7 1.64q.54 0 .938.203.405.203.623.567.224.364.224.854 0 .483-.224.854-.224.364-.623.567a2.04 2.04 0 0 1-.938.203h-.693V19h-1.05v-5.11zm4.302 0q.533 0 .924.203.4.203.616.574.224.365.224.86v1.828q0 .49-.224.86-.216.372-.616.582-.391.203-.924.203h-1.659v-5.11zm2.798 0V19h1.05v-2.016h2.058v-.98h-2.072V14.87h2.254v-.98zm-6.568 1.134q-.196-.19-.532-.19h-.693v1.359H7.7q.336 0 .532-.182a.65.65 0 0 0 .203-.497.63.63 0 0 0-.203-.49m4.288.028a.73.73 0 0 0-.518-.182h-.609v3.15h.61q.322 0 .517-.182a.63.63 0 0 0 .196-.483v-1.827a.62.62 0 0 0-.196-.476"
      clip-rule="evenodd"
    >
    </path>
  </svg>
  Samisk i barnehagen (PDF, 299KB)
</a>
`;

exports[`Focused 1`] = `
<a href="https://udir.no/">
  Gå til udir.no
</a>
`;

exports[`In Text 1`] = `
<p data-variant="default">
  Designsystemet må ses i sammenheng med
  <a href="https://www.udir.no/om-udir/mal-og-oppgaver/">
    Udirs mål og oppgaver
  </a>
  .
</p>
`;

exports[`Long Link 1`] = `
<p data-variant="default">
  <a href="https://udir.no/">
    Dette er en lenke som brekker over flere linjer
  </a>
</p>
`;

exports[`Preview 1`] = `
<a href="https://udir.no/">
  Gå til udir.no
</a>
`;

exports[`With Icon 1`] = `
<a href="mailto:designsystem@digdir.no">
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="1em"
    height="1em"
    fill="none"
    viewbox="0 0 24 24"
    focusable="false"
    role="img"
    aria-hidden="true"
  >
    <path
      fill="currentColor"
      fill-rule="evenodd"
      d="M3 5.25a.75.75 0 0 0-.75.75v13c0 .414.336.75.75.75h18a.75.75 0 0 0 .75-.75V6a.75.75 0 0 0-.75-.75zm.75 11.94 5.273-5.273L3.75 8.4zm6.546-4.425L4.81 18.25h14.378l-5.485-5.485-1.288.859a.75.75 0 0 1-.832 0zm4.681-.848L20.25 8.4v8.788zm-11-5.167L12 12.099l8.023-5.349z"
      clip-rule="evenodd"
    >
    </path>
  </svg>
  Kontakt oss
</a>
`;
